app/services/change_schedule.rb

```diff
@@ -0,0 +1,187 @@
+      participant_profile.update_schedule!(schedule)
+
+      if relevant_induction_record
+        Induction::ChangeInductionRecord.call(
+          induction_record: relevant_induction_record,
+          changes: {
+            schedule:,
+          },
+        )
+      end
+    end
+
+    participant_profile.record_to_serialize_for(lead_provider: cpd_lead_provider.lead_provider)
+  end
+
+  def participant_identity
+    @participant_identity ||= ParticipantIdentity.find_by(external_identifier: participant_id)
+  end
+
+  def participant_profile
+    @participant_profile ||= ParticipantProfileResolver
+                               .call(
+                                 participant_identity:,
+                                 course_identifier:,
+                                 cpd_lead_provider:,
+                               )
+  end
+
+  def schedule
+    participant_profile&.schedule_for(cpd_lead_provider:)
+  end
+
+  def alias_search_query
+    Finance::Schedule
+      .where("identifier_alias IS NOT NULL")
+      .where(identifier_alias: schedule_identifier, cohort:)
+  end
+
+  def schedule
+    @schedule ||= Finance::Schedule
+      .where(schedule_identifier:, cohort:)
+      .or(alias_search_query)
+      .first
+  end
+
+private
+
+  def user
+    @user ||= participant_identity&.user
+  end
+
+  def cohort
+    @cohort ||= if super
+                  Cohort.find_by(start_year: super)
+                else
+                  Cohort.current
+                end
+  end
+
+  def relevant_induction_record
+    return if user.blank? || participant_profile.blank?
+
+    @relevant_induction_record ||= participant_profile.latest_induction_record_for(cpd_lead_provider:)
+  end
+
+  def not_already_withdrawn
+    return unless participant_profile
+
+    errors.add(:participant_id, I18n.t(:withdrawn_participant)) if participant_profile.withdrawn_for?(cpd_lead_provider:)
+  end
+
+  def validate_new_schedule_valid_with_existing_declarations
+    return if user.blank? || participant_profile.blank?
+    return unless schedule
+
+    participant_profile.participant_declarations.each do |declaration|
+      next unless %w[submitted eligible payable paid].include?(declaration.state)
+
+      milestone = schedule.milestones.find_by!(declaration_type: declaration.declaration_type)
+
+      if declaration.declaration_date <= milestone.start_date.beginning_of_day
+        errors.add(:schedule_identifier, I18n.t(:schedule_invalidates_declaration))
+      end
+
+      if milestone.milestone_date && (milestone.milestone_date.end_of_day < declaration.declaration_date)
+        errors.add(:schedule_identifier, I18n.t(:schedule_invalidates_declaration))
+      end
+    end
+  end
+
+  def validate_provider
+    return if user.blank? || participant_profile.blank?
+
+    unless participant_profile && matches_lead_provider?
+      errors.add(:participant_id, I18n.t(:invalid_participant))
+    end
+  end
+
+  def matches_lead_provider?
```

Comment:
maybe we can add a shared interface for NPQ/ECF profiles similar to withdrawn etc.?

Reply:
@cwrw All done tks